File: course1_algorithmic_toolbox/week2_algorithmic_warmup/8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/FibonacciSumSquares.java
import java.util.Scanner;

public class FibonacciSumSquares {

	private static long getFibonacciSumSquares(long n) {
		if (n <= 1) {
			return n;
		}

		int[] firstSixtyOfLastDigits = firstSixtyLastDigits();
		long quotient = n / 60;
		int remain = (int) (n % 60);

		return (quotient * sumSquares(firstSixtyOfLastDigits) + sumSquaresUntil(firstSixtyOfLastDigits, remain)) % 10;
	}

	/**
	 * Sun squares of an array until the given index
	 *
	 * @param digits
	 * @param until
	 * @return
	 */
	private static int sumSquaresUntil(int[] digits, int until) {
		if (until >= digits.length) {
			throw new RuntimeException("i can not be greater than or equal " + digits.length);
		}

		int sum = 0;
		for (int i = 0; i <= until; i++) {
			sum += digits[i] * digits[i];
		}

		return sum;
	}

	/**
	 * Sum squares of an array
	 *
	 * @param digits
	 * @return
	 */
	private static int sumSquares(int[] digits) {
		int sum = 0;
		for (int digit : digits) {
			sum += digit * digit;
		}

		return sum;
	}

	/**
	 * The last digits repeat every 60 numbers.
	 *
	 * @return
	 */
	private static int[] firstSixtyLastDigits() {
		int[] digits = new int[60];
		digits[0] = 0;
		digits[1] = 1;

		for (int i = 2; i < 60; i++) {
            digits[i] = (digits[i - 1] + digits[i - 2]) % 10;
		}

		return digits;
	}

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		long n = scanner.nextLong();
		long s = getFibonacciSumSquares(n);
		System.out.println(s);
	}
}

File: course1_algorithmic_toolbox/week2_algorithmic_warmup/8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/fibonacci_sum_squares.js
const readline = require('readline');
const rl = readline.createInterface({
	input: process.stdin,
	terminal: false
});
process.stdin.setEncoding('utf8');

/**
 * Sun squares of an array until the given index
 *
 * @param digits
 * @param until
 * @returns {number}
 */
const sumSquaresUntil = (digits, until) => {
	if (until >= digits.length) {
		throw new Error("i can not be greater than or equal " + digits.length);
	}

	let sum = 0;
	for (let i = 0; i <= until; i++) {
		sum += digits[i] * digits[i];
	}

	return sum;
};

/**
 * Sum square of an array
 *
 * @param digits
 * @returns {number}
 */
const sumSquares = (digits) => {
	let sum = 0;
	digits.forEach(digit => sum += digit * digit);
	return sum;
};

/**
 * The last digits repeat every 60 numbers.
 */
const firstSixtyOfLastDigits = () => {
	const digits = [];
	digits[0] = 0;
	digits[1] = 1;

	for (let i = 2; i < 60; i++) {
		digits[i] = (digits[i - 1] + digits[i - 2]) % 10;
	}

	return digits;
};

/**
 * Get Fibonacci sum of last digits from F0 to Fn
 *
 * @param n
 * @returns {number|*}
 */
const getFibonacciSumSquares = (n) => {
	if (n <= 1) {
		return n;
	}

	const digits = firstSixtyOfLastDigits();
	const quotient = Math.floor(n / 60);
	const remain = n % 60;

	return (quotient * sumSquares(digits) + sumSquaresUntil(digits, remain)) % 10;

};

const readLine = (line) => {
	console.log(getFibonacciSumSquares(parseInt(line, 10)));
	process.exit();
};


rl.on('line', readLine);

module.exports = getFibonacciSumSquares;
